knative-operator/pkg/monitoring/sources/source_deployment_discovery_controller.go

```diff
@@ -69,38 +79,133 @@ type ReconcileSourceDeployment struct {
 	// that reads objects from the cache and writes to the apiserver
 	client client.Client
 	scheme *runtime.Scheme
+
+	testShouldUseClusterMonitoring          *bool
+	testShouldGenerateSourceServiceMonitors *bool
 }
 
 // Reconcile reads that state of the cluster for an eventing source deployment
 func (r *ReconcileSourceDeployment) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
 	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
 	reqLogger.Info("Reconciling the source deployment, setting up a service/service monitor if required")
 	dep := &v1.Deployment{}
-	if err := r.client.Get(context.TODO(), request.NamespacedName, dep); err != nil {
-		// the deployment does not exist anymore do nothing
-		return reconcile.Result{}, nil
-	}
-	if err := monitoring.SetupMonitoringRequirements(r.client, dep); err != nil {
+	err := r.client.Get(context.TODO(), request.NamespacedName, dep)
+	inDeletion := false
+	if apierrors.IsNotFound(err) {
+		// The deployment does not exist anymore, deletions are shown as failing reads
+		log.Info("Source in deletion phase")
+		inDeletion = true
+	} else if err != nil {
 		return reconcile.Result{}, err
 	}
-	if err := SetupSourceServiceMonitorResources(r.client, dep); err != nil {
+	eventing := &v1alpha1.KnativeEventing{}
+	if err = r.client.Get(context.TODO(), types.NamespacedName{Namespace: "knative-eventing", Name: "knative-eventing"}, eventing); err != nil {
 		return reconcile.Result{}, err
 	}
-	return reconcile.Result{}, nil
+	shouldEnableMonitoring := okomon.ShouldEnableMonitoring(eventing.Spec.GetConfig())
+	return reconcile.Result{}, r.reconcileSourceMonitoring(dep, shouldEnableMonitoring, inDeletion)
```

Comment:
If someone disables monitoring eg. set backend to `none` this triggers an update to all source adapter deployments.

Reply:
> Shouldn't inDeletion be triggered by a finalizer being present? If we need to cleanup arbitrary (non ownerRef) stuff from the Deployment, we should be adding a finalizer to it. That would cause this to no longer work though, as it'd stick around until the finalizer has been removed by us.

If we were to manage the non ownerRefresources like Prometheus rbac rules yes. Currently if someone accidentally deletes them then I dont restore them unless some source update happens. Otherwise will have to setup a watcher/reconciler for any different kind involved.